httpbp/client_middlewares.go

```diff
@@ -0,0 +1,239 @@
+// plus any additional client middleware passed into this function. Default
+// middlewares are: MonitorClient and Retries. ClientErrorWrapper is included
+// as transitive middleware through Retries.
+func NewClient(config ClientConfig, middleware ...ClientMiddleware) (*http.Client, error) {
+	if err := config.Validate(); err != nil {
+		return nil, err
+	}
+
+	// set max connections per host if set
+	var transport http.Transport
+	if config.MaxConnections > 0 {
+		transport.MaxConnsPerHost = config.MaxConnections
+	}
+
+	// apply default if not set
+	if config.MaxErrorReadAhead == 0 {
+		config.MaxErrorReadAhead = DefaultMaxErrorReadAhead
+	}
+
+	// if no retry options are set default to retry.Attempts(1)
+	if len(config.RetryOptions) == 0 {
+		config.RetryOptions = []retry.Option{retry.Attempts(1)}
+	}
+
+	defaults := []ClientMiddleware{
+		MonitorClient(config.Slug),
+		Retries(config.MaxErrorReadAhead, config.RetryOptions...),
+	}
+
+	// prepend middleware to ensure Retires with ClientErrorWrapper is still
+	// applied first
+	if config.CircuitBreaker != nil {
+		defaults = append([]ClientMiddleware{CircuitBreaker(*config.CircuitBreaker)}, defaults...)
+	}
+	middleware = append(middleware, defaults...)
+
+	return &http.Client{
+		Transport: WrapTransport(&transport, middleware...),
+	}, nil
+}
+
+// WrapTransport takes a list of client middleware and wraps them around the
+// given transport. This is useful for using client middleware outside of this
+// package.
+func WrapTransport(transport http.RoundTripper, middleware ...ClientMiddleware) http.RoundTripper {
+	if transport == nil {
+		transport = http.DefaultTransport
+	}
+	// add middleware in reverse so the first in the list is the outermost
+	for i := len(middleware) - 1; i >= 0; i-- {
+		transport = middleware[i](transport)
+	}
+	return transport
+}
+
+// ClientErrorWrapper applies ClientErrorFromResponse to the returned response
+// ensuring an HTTP status response outside the range [200, 400) is wrapped in
+// an error relieving users from the need to check the status response.
+//
+// If a response is wrapped in an error this middleware will perform
+// DrainAndClose on the response body and will read up to limit to store
+// additional info about the HTTP response.
+func ClientErrorWrapper(limit int) ClientMiddleware {
+	return func(next http.RoundTripper) http.RoundTripper {
+		return roundTripperFunc(func(req *http.Request) (resp *http.Response, err error) {
+			resp, err = next.RoundTrip(req)
+			if err != nil {
+				return resp, err
+			}
+			if err == nil {
+				err = ClientErrorFromResponse(resp)
+				if err != nil {
+					defer DrainAndClose(resp.Body)
+					var ce *ClientError
+					if !errors.As(err, &ce) {
+						return resp, err
+					}
+					body, e := io.ReadAll(io.LimitReader(resp.Body, int64(limit)))
+					if e != nil {
+						return resp, e
+					}
+					ce.AdditionalInfo = string(body)
+					return resp, ce
```

Comment:
we need to change `resp` to nil on lines 112, 116, 119, and do

```go
if err != nil {
  return nil, err
}
return resp, nil
```

on line 122.